File: word_square.py


class Solution(object):
    def wordSquares(self, words):
        """
        :type words: List[str]
        :rtype: List[List[str]]
        """
        self.words = words
        self.N = len(words[0])
        self.buildTrie(self.words)

        res = []
        word_squares = []
        
        for word in words:
            word_squares = [word]
            self.backtracking(1, word_squares, res)
        return res
    
    # added a new function to build a Trie out of the input words.
    def buildTrie(self, words):
        self.trie = {}

        for wordIndex, word in enumerate(words):
            node = self.trie
            for char in word:
                if char in node:
                    node = node[char]
                else:
                    newNode = {}
                    newNode['#'] = []
                    node[char] = newNode
                    node = newNode
                node['#'].append(wordIndex)

    def backtracking(self, step, word_squares, res):
        if step == self.N:
            res.append(word_squares[:])
            return

        prefix = ''.join([word[step] for word in word_squares])
        for candidate in self.getWordsWithPrefix(prefix):
            word_squares.append(candidate)
            self.backtracking(step+1, word_squares, res)
            word_squares.pop()
    
    #query the Trie to retrieve all the words that possess the given prefix.
    def getWordsWithPrefix(self, prefix):
        node = self.trie
        for char in prefix:
            if char not in node:
                return []
            node = node[char]
        return [self.words[wordIndex] for wordIndex in node['#']]

File: wordSquare.java
class Solution {
    public List<List<String>> wordSquares(String[] words) {
        List<List<String>> result = new ArrayList<>();
        if (words == null || words.length == 0) {
            return result;
        }
        
        int squareSize = words[0].length();
        TrieNode root = buildTrie(words);
        
        List<String> square = new ArrayList<>();
        for (String word : words) {
            square.add(word);
            backtrack(result, square, squareSize, root);
            square.remove(square.size() - 1);
        }
        
        return result;
    }
    
    private TrieNode buildTrie(String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int index = c - 'a';
                if (node.children[index] == null) {
                    node.children[index] = new TrieNode();
                }
                node = node.children[index];
                node.words.add(word);
            }
        }
        return root;
    }
    
    private void backtrack(List<List<String>> result, List<String> square, int squareSize, TrieNode root) {
        if (square.size() == squareSize) {
            result.add(new ArrayList<>(square));
            return;
        }
        
        int index = square.size();
        StringBuilder prefix = new StringBuilder();
        for (String word : square) {
            prefix.append(word.charAt(index));
        }
        
        List<String> candidates = getCandidates(prefix.toString(), root);
        for (String candidate : candidates) {
            square.add(candidate);
            backtrack(result, square, squareSize, root);
            square.remove(square.size() - 1);
        }
    }
    
    private List<String> getCandidates(String prefix, TrieNode root) {
        List<String> candidates = new ArrayList<>();
        TrieNode node = root;
        for (char c : prefix.toCharArray()) {
            int index = c - 'a';
            if (node.children[index] == null) {
                return candidates;
            }
            node = node.children[index];
        }
        candidates.addAll(node.words);
        return candidates;
    }
    
    class TrieNode {
        TrieNode[] children;
        List<String> words;
        
        TrieNode() {
            children = new TrieNode[26];
            words = new ArrayList<>();
        }
    }
}

